concurrency/MemDBDomainManager.java

```diff
@@ -0,0 +1,59 @@
+package concurrency;
+
+import com.amazon.aws.ccs.persistence.DomainManager;
+import com.amazon.aws.ccs.persistence.exceptions.PersistenceException;
+import com.amazon.coral.metrics.Metrics;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeoutException;
+
+public class MemDBDomainManager implements DomainManager<MemDBDomain> {
+
+  private final Map<String, MemDBDomain> domains;
+  private final String namespace;
+
+  public MemDBDomainManager(String namespace) {
+    this.namespace = namespace;
+    this.domains = new ConcurrentHashMap<>();
+
+  }
+
+  @Override
+  public MemDBDomain create(MemDBDomain domain, boolean waitUntilReady, boolean failIfExists) throws PersistenceException,
+      TimeoutException {
+
+    this.domains.put(domain.getName(), domain);
+
+    return domain;
+  }
+
+  @Override
+  public void configureDomain(MemDBDomain domain, Map<String, String> properties) throws TimeoutException {
+    // noop
+  }
+
+  @Override
+  public synchronized MemDBDomain retrieve(String domainName) throws PersistenceException {
+    return this.domains.get(domainName);
+  }
+
+  @Override
+  public synchronized void delete(String domainName) throws PersistenceException {
+    MemDBDomain domainToDelete = this.domains.get(domainName);
+    if (domainToDelete != null) {
+      this.domains.remove(domainName);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 These lines use `ConcurrentHashMap`&#39;s **sequential** API and are protected by `synchronized` to ensure they execute together: **43 and 45**. This synchronization may reduce performance by limiting parallelism and does not take advantage of `ConcurrentHashMap`&#39;s **concurrent** API.

**Fix**  
 Consider replacing the check and `remove(key)` with `remove(key, value)` and then removing the synchronization. The check is at line: **44**. If you do not need the check, consider removing the check and the synchronization and just calling `remove(key)`.

**More info**  
 `remove(key, value)` removes the key only if the current value is equal to the `value` parameter. `remove(key, value)` returns `true` if the call removed the `key` and `value`.

 [View an example on GitHub](https://github.com/apache/hadoop/blob/4d3c580b03475a6ec9323d11e6875c542f8e3f6d/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/ipc/Client.java#L1591-L1592) (external link).